### PR TITLE
Runner: Listen for context cancellation and exit liveness listener

### PR DESCRIPTION
**Pull request by catsby**
Gracefully quit the liveness listener if accepting a job errors and the context gets cancelled. Before this change, the liveness server would continue to operate and spew errors in the logs.

Example based off my findings from debugging  https://github.com/hashicorp/waypoint/issues/1398:

<details>

Before:

```
2021-06-24T17:58:03.538Z [INFO]  waypoint.runner.agent: initializing the runner
2021-06-24T17:58:03.538Z [INFO]  waypoint.runner.agent: starting runner
2021-06-24T17:58:03.538Z [DEBUG] waypoint.runner.agent.runner: registering runner
2021-06-24T17:58:03.539Z [TRACE] waypoint.runner.agent.runner: runner connected, waiting for initial config
2021-06-24T17:58:03.541Z [INFO]  waypoint.runner.agent.runner: runner registered with server
2021-06-24T17:58:03.541Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-06-24T17:58:03.541Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-06-24T17:58:03.542Z [INFO]  waypoint.runner.agent.runner: waiting for job assignment
2021-06-24T17:59:02.612Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-24T17:59:02.612Z [WARN]  waypoint.runner.agent: server unavailable, sleeping before retry
2021-06-24T17:59:02.612Z [ERROR] waypoint.runner.agent.runner.config_recv: error receiving configuration, exiting: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-24T17:59:02.612Z [TRACE] waypoint.runner.agent.runner.config_recv: exiting receive goroutine
2021-06-24T17:59:04.612Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-06-24T17:59:04.613Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-06-24T17:59:04.613Z [INFO]  waypoint.runner.agent.runner: waiting for job assignment
2021-06-24T17:59:04.615Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = NotFound desc = runner ID not found"
2021-06-24T17:59:04.615Z [ERROR] waypoint.runner.agent: runner unexpectedly deregistered, exiting
2021-06-24T17:59:04.615Z [INFO]  waypoint.runner.agent: quit request received, gracefully stopping runner
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-24T17:59:04.615Z [TRACE] waypoint: stopping signal listeners and cancelling the context
2021-06-24T17:59:04.615Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
```

After:

```
2021-06-28T21:16:31.202Z [INFO]  waypoint.runner.agent: initializing the runner
2021-06-28T21:16:31.202Z [INFO]  waypoint.runner.agent: starting runner
2021-06-28T21:16:31.202Z [DEBUG] waypoint.runner.agent.runner: registering runner
2021-06-28T21:16:31.202Z [TRACE] waypoint.runner.agent.runner: runner connected, waiting for initial config
2021-06-28T21:16:31.206Z [INFO]  waypoint.runner.agent.runner: runner registered with server
2021-06-28T21:16:31.208Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-06-28T21:16:31.208Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-06-28T21:16:31.209Z [INFO]  waypoint.runner.agent.runner: waiting for job assignment
2021-06-28T21:17:30.669Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-28T21:17:30.669Z [WARN]  waypoint.runner.agent: server unavailable, sleeping before retry
2021-06-28T21:17:30.670Z [ERROR] waypoint.runner.agent.runner.config_recv: error receiving configuration, exiting: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-28T21:17:30.670Z [TRACE] waypoint.runner.agent.runner.config_recv: exiting receive goroutine
2021-06-28T21:17:32.671Z [DEBUG] waypoint.runner.agent.runner: opening job stream
2021-06-28T21:17:32.671Z [TRACE] waypoint.runner.agent.runner: sending job request
2021-06-28T21:17:32.671Z [INFO]  waypoint.runner.agent.runner: waiting for job assignment
2021-06-28T21:17:32.673Z [ERROR] waypoint.runner.agent: error running job: err="rpc error: code = NotFound desc = runner ID not found"
2021-06-28T21:17:32.673Z [ERROR] waypoint.runner.agent: runner unexpectedly deregistered, exiting
2021-06-28T21:17:32.673Z [INFO]  waypoint.runner.agent: quit request received, gracefully stopping runner
2021-06-28T21:17:32.673Z [WARN]  waypoint.runner.agent: error accepting liveness connection: err="accept tcp [::]:1234: use of closed network connection"
2021-06-28T21:17:32.673Z [TRACE] waypoint: stopping signal listeners and cancelling the context

```

</details>

### CODE REVIEW

File: internal/cli/runner_agent.go [hunk 1]
```diff
@@ -125,12 +129,19 @@ func (c *RunnerAgentCommand) Run(args []string) int {
 				"Error starting liveness server: %s", err.Error(),
 				terminal.WithErrorStyle(),
 			)
+			cancel()
```
Comment (mitchellh) [inline internal/cli/runner_agent.go:132]: Doesn't the `return 1` below exit us already? So we're canceling something that'll never be run, aren't we?

**Review by izaaklauer — APPROVED** [review]
LGTM. Looks to me like that context only cancels if the runner Accept loop fails, and in that case we shouldn't orphan the liveness check goroutine.

File: internal/cli/runner_agent.go (hunk 1 above)
Comment (catsby) [reply internal/cli/runner_agent.go]: I wasn't 100% sure on if the `cancel` was needed here, but I imagine you're correct and it's not needed. I removed it in 74e7982a

**Comment by mitchellh** [comment]
I'm still a bit confused (and not entirely convinced) this fixes the linked issue.

The linked issue is crashing cause `conn` is nil. I have no idea how conn becomes nil if `Accept` returns with a nil error (the only way we reach the crashing line). Maybe you've seen something I don't?

I don't think we need context cancellation where you've put it, and even if we do, its still racy because you can get stuck blocking on `ln.Accept` while the context is NOT canceled. My thinking around this was: we dont ever need to stop this goroutine because process exit will kill everything anyways.

**Comment by catsby** [comment]
@mitchellh I simplified this to just assert the error and leave the context as it was, let me know what you think

**Comment by mitchellh** [comment]
Oh yeah that’s super clean. Good work!

**Pull request merged**